### DISCUSSION TITLE
Understanding enum EpdDrawMode

### DISCUSSION THREAD

**Discussion by holgerlembke** [vroland/epdiy#331] [General]
Moin.

I have a ED097TC2.

After epd_clear(); I get a nice white background. Drawing some lines and then doing
```
EpdDrawError err = epd_hl_update_screen(&epdiydata.hl, MODE_GC16, temperature);
```
gives the line (not very crisp) and very lightly gray areas. Doing it with 

```
EpdDrawError err = epd_hl_update_screen(&epdiydata.hl, MODE_GL16, temperature);
```
gives a way better result, but lines fade out in the outer areas of the display.

**Looks like I don't understand what this thing does and what to choose.**

**Comment by martinberlin** [comment]
Those are two different 16 grays modes.
If you only want to draw monochrome I recommend to use faster MODE_DU.
You can read more about Eink in general here:
https://github.com/Modos-Labs/Glider

Is a well documented repository from Wenting Zhang that knows quite a lot of this displays.

basically this modes are different timings to go from a gray to another gray using 16 tones.
This are stored in waveform files and each display configuration uses one of them. They are not resolution dependent they are just different timings to go from one gray to another and so for for me is the hardest part of this project.

**Comment by holgerlembke** [comment]
I looked into it.

What I don't understand is that with MODE_GC16 I get areas of slightly gray scale. Not all, not everywhere.

MODE_DU/MODE_GL16 gives the same visual results to me. I do not understand the edges painting.

![grafik](https://github.com/vroland/epdiy/assets/2034993/6e47891a-af8e-4b52-bc86-fd3a85c0252a)

And for my examples, what should I write as an advice for user? "use MODE_GC16 if you want to display stuff with gray scales. use MODE_DU if you draw with black/white lines/text."?

**Reply by martinberlin** [reply]
Oh that's weird. Should not happen. It does not happen with my display.
Can you try to draw a filled rectangle of the whole size of the display?

Yes you should use GC16 only if you need grayscales. Only monochrome will be much faster and efficient, also more less consumption since uses less cycles to use MODE_DU

**Comment by holgerlembke** [comment]
```
EpdRect rect = {
    .x = 0,
    .y = 0,
    .width = epd_width(),
    .height = epd_height()
  };
  epd_fill_rect(rect,black,epdiydata.epaperFrameBuffer);
```
gives a nice black panel. No visual problems.

**Comment by martinberlin** [comment]
Please copy the C code to get that crossed drawing and I will try to draw exactly the same with v7 board

**Comment by holgerlembke** [comment]
https://github.com/holgerlembke/epdiy/tree/main/examples/Basics/A_First_Steps

(second try... :-)

**Comment by martinberlin** [comment]
![AADA2901-737F-4B5E-978A-2E3E0BBD8924](https://github.com/vroland/epdiy/assets/2692928/2ec4b269-5356-4951-90b3-9362311d26fb)

Bad light here… but I also see some gray areas at the side of the circle. @vroland do you have any idea why drawing this makes such an effect?

**Comment by holgerlembke** [comment]
Looks exactly as mine...

JPEG-Downloader is working fine. I'm not sooo confident in the stuff the jpeg decoder does with the dithering. I need to look into that.

![grafik](https://github.com/vroland/epdiy/assets/2034993/921390fe-4191-4953-859b-8b74f8d2cb35)

**Reply by martinberlin** [reply]
That looks like a healthy feline

**Reply by holgerlembke** [reply]
not mine. is from cats as a service, https://cataas.com/cat
